pkg/deployments/scaler.go

```diff
@@ -46,69 +70,53 @@ func (s *scaler) UpdateState(replicas, min, max int32) {
 // SetDesiredScale sets the desired scale of the scaler and scales
 // if needed.
 func (s *scaler) SetDesiredScale(n int32) {
-	log.Printf("SetDesiredScale(%v)", n)
-	s.compareScales(-1, s.applyMinMax(n))
-}
-
-func (s *scaler) applyMinMax(n int32) int32 {
-	s.mtx.Lock()
-	min := s.minScale
-	max := s.maxScale
-	s.mtx.Unlock()
-	if n < min {
-		n = min
-	} else if n > max {
-		n = max
-	}
-	return n
-}
-
-func (s *scaler) compareScales(current, desired int32) {
+	log.Printf("SetDesiredScale(%v), current: %v, min: %v, max: %v", n, s.currentScale, s.minScale, s.maxScale)
+	nMinMax := s.applyMinMax(n)
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
-
-	if current != -1 {
-		s.currentScale = current
-	}
-	if desired != -1 {
-		s.desiredScale = desired
-	}
-
-	if s.currentScale == -1 || s.desiredScale == -1 {
-		// Nothing to compare if we only have partial information
-		return
-	}
-
-	log.Printf("Comparing scales, current: %v, desired: %v", s.currentScale, s.desiredScale)
+	s.desiredScale = nMinMax
 
 	if s.desiredScale > s.currentScale {
 		// Scale up immediately.
-		go s.scaleFunc(s.desiredScale, false)
-		s.scaleDownStarted = false
+		log.Printf("Scaling up immediately")
+		go func() {
+			if err := s.scaleFunc(s.desiredScale, false); err != nil {
+				log.Printf("scale down to %v error: %v", s.desiredScale, err)
+			} else {
+				log.Printf("Scaled up to %v successfully", s.desiredScale)
+			}
+		}()
 	} else if s.desiredScale == s.currentScale {
-		// Do nothing, schedule nothing.
-		if s.scaleDownTimer != nil {
-			s.scaleDownTimer.Stop()
-		}
-		s.scaleDownStarted = false
+		log.Println("Desired scale equals current scale, doing nothing")
 	} else {
-		// Schedule a scale down.
-
-		if s.scaleDownTimer == nil {
-			s.scaleDownTimer = time.AfterFunc(s.scaleDownDelay, func() {
+		if time.Since(s.lastScaleDown) >= s.scaleDownDelay {
```

Comment:
🤔 the `lastScaleDown` value is set in the constructor. This ensures the scale down delay is taken into account on a node startup and right after the last successful scale down. 
There is no delay if a running node becomes leader or if the last scale down is some time ago.  A scale down call would execute immediately as the condition is always true.
With the old logic, the first call to scale down triggered a timer. I order to achieve the same, we can have an unset `lastScaleDown` time and set it on first call to scale down. You can check with `s.lastScaleDown.IsZero()` for unset.
It should be cleared (unset) on 
* desiredScale >= currentScale
* successful scale down
* loosing leader role
* Probably with `AtLeastOne()`

Reply:
Here is my code I was working on in parallel:

```go
	if s.currentScale < s.desiredScale {
		// Scale up immediately.
		go s.scaleFunc(s.desiredScale, false)
		s.desiredScaleDownStart = time.Time{}
	} else if s.currentScale == s.desiredScale {
		s.desiredScaleDownStart = time.Time{}
	} else {
		if s.desiredScaleDownStart.IsZero() {
			s.desiredScaleDownStart = time.Now()
		} else if time.Since(s.desiredScaleDownStart) >= s.scaleDownDelay {
			go s.scaleFunc(s.desiredScale, false)
		}
	}
```

Reply:
Thanks for taking the time to do a deeper analysis on how it differs. I agree that setting it to zero value is cleaner and have fixed it.

@nstogner regarding your implementation, it's important that we always check and log the error of s.scaleFunc because otherwise it will be hard to debug if there is a scaling issue. Was there any reason we weren't originally doing error checks? and any reason why your implementation does not? Just making sure I'm not doing something useless with the error checking.